python/cog/server/worker.py

```diff
@@ -171,18 +178,30 @@ def run(self) -> None:
             [ws_stdout, ws_stderr], self._stream_write_hook
         )
         self._stream_redirector.start()
-
         self._setup()
-        asyncio.run(self._loop())
+        self._loop()
```

Comment:
Another option that might (?) make things less icky, if I've finally understood the problem here.

```python
def run(self) -> None:
    ...
 
    self._stream_redirector.start()
    asyncio.run(self._run())
    self._stream_redirector.shutdown()

async def _run(self) -> Awaitable[None]:
    await self._setup()
    await self._loop()

async def _setup(self) -> Awaitable[None]:
    ...
```

Reply:
we don't know if we're async or not until we load_predictor_from_ref, which might throw an error, that should be handled in the same block as errors from calling setup

Reply:
we could duplicate the setup error handling with a similar pattern as _handle_predict_error and then this could work, I was just a little confused what was happening with Done and _cancelable

Reply:
Maybe I'm missing something, but why does the predictor need to be async for these methods to be async?

Reply:
if the predictor isn't async, they haven't opted into the async world, and if we make these methods async we would get the same problems we were having before with there being a running event loop that the user isn't expecting, breaking unsuspecting code

Reply:
we could have a kind of silly check like `"async" in open("predict.py").read()`